aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,173 @@
+
+.. note:: Returning anonymous types is a common practice, especially in web API methods, since it's a simple, easy way to control the shape of the output. However, anonymous types are ``internal``, so unit tests that attempt to refer to their properties may encounter errors unless the web project is configured to make its internals visible to the test project.
+
+The unit tests for the ``ForSession`` API:
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/UnitTests/ApiIdeasControllerForSession.cs
+  :language: c#
+  :emphasize-lines: 14-15,25-26,36-37
+
+The second test, ``ReturnsIdeasForSession``, initially fails with this message:
+
+.. code-block: c#
+
+'object' does not contain a definition for 'name'
+
+To correct this error, add an assembly directive specifying that the web project's internals should be visible to the test client. This directive can go into its own file (``InternalsVisibleTo.cs``):
+
+.. literalinclude:: testing/sample/TestingControllersSample/src/TestingControllersSample/InternalsVisibleTo.cs
+  :language: c#
+  :emphasize-lines: 3
+
+With this in place, the test will pass, since the unit test project will have access to the internal-scoped anonymous return type. This is only a problem for unit tests that directly inspect the result - integration tests do not require ``InternalsVisibleTo`` since they get a result from an ``HttpClient``, not a direct method return.
+
+The unit tests for the ``Create``:
+
+.. literalinclude:: testing/sample/TestingControllersSample/tests/TestingControllerSample.Tests/UnitTests/ApiIdeasControllerCreate.cs
+  :language: c#
+  :emphasize-lines: 13-14,18,23-24,28,35-36
+
+As stated previously, to test the behavior of the method when ``ModelState`` is invalid, add a model error to the controller. Don't try to test model validation or model binding in your unit tests - just test your action method's behavior.
+
+The second test depends on the repository returning null, so the mock repository is configured to return null. There's no need to create a test database (in memory or otherwise) and construct a query that will return this result - it can be done in a single line as shown.
+
+The last test verifies that the repository's ``Update`` method is called. As we did previously, the mock is called with ``Verifiable`` and then the mocked repository's ``Verify`` method is called to confirm the verifiable method was executed. It's not a unit test responsibility to ensure that the ``Update`` method saved the data; that should be done with an integration test.
+
+Integration Testing
```

Comment:
I think this section does a good job of documenting ASP.NET Core specific concerns for integration testing. Since integration testing is generally not controller specific I wonder if this content should move to https://docs.asp.net/en/latest/testing/integration-testing.html. Should we spit this article in two and move them both to the top level Testing node?